src/vmm/src/vstate/vcpu/x86_64.rs

```diff
@@ -252,35 +273,50 @@ impl KvmVcpu {
         guest_mem: &GuestMemoryMmap,
         kernel_start_addr: GuestAddress,
         vcpu_config: &VcpuConfig,
-        mut cpuid: CpuId,
+        cpuid: CpuId,
     ) -> std::result::Result<(), KvmVcpuConfigureError> {
-        let cpuid_vm_spec = VmSpec::new(self.index, vcpu_config.vcpu_count, vcpu_config.smt)
-            .map_err(KvmVcpuConfigureError::VmSpec)?;
-
-        filter_cpuid(&mut cpuid, &cpuid_vm_spec)
-            .map_err(|err| {
-                METRICS.vcpu.filter_cpuid.inc();
-                error!(
-                    "Failure in configuring CPUID for vcpu {}: {:?}",
-                    self.index, err
-                );
-                err
-            })
-            .map_err(KvmVcpuConfigureError::FilterCpuid)?;
-
-        // Update the CPUID based on the template
-        match vcpu_config.cpu_template {
-            CpuFeaturesTemplate::T2 => t2::set_cpuid_entries(&mut cpuid, &cpuid_vm_spec)
-                .map_err(KvmVcpuConfigureError::SetCpuidEntries)?,
-            CpuFeaturesTemplate::T2S => t2s::set_cpuid_entries(&mut cpuid, &cpuid_vm_spec)
-                .map_err(KvmVcpuConfigureError::SetCpuidEntries)?,
-            CpuFeaturesTemplate::C3 => c3::set_cpuid_entries(&mut cpuid, &cpuid_vm_spec)
-                .map_err(KvmVcpuConfigureError::SetCpuidEntries)?,
-            CpuFeaturesTemplate::None => {}
+        // If a template is specified, get the CPUID template, else use `cpuid`.
+        let mut config_cpuid = match vcpu_config.cpu_template {
+            CpuFeaturesTemplate::T2 => cpuid_templates::t2(),
+            CpuFeaturesTemplate::T2S => cpuid_templates::t2s(),
+            CpuFeaturesTemplate::C3 => cpuid_templates::c3(),
+            // If a template is not supplied we use the given `cpuid` as the base.
+            CpuFeaturesTemplate::None => cpuid::Cpuid::try_from(cpuid::RawCpuid::from(cpuid))
+                .map_err(KvmVcpuConfigureError::SnapshotCpuid)?,
+        };
+
+        // Apply adjustments based on `vm_spec`.
+        {
+            config_cpuid
+                .normalize(
+                    // The index of the current logical CPU in the range [0..cpu_count].
+                    self.index,
+                    // The total number of logical CPUs.
+                    vcpu_config.vcpu_count,
+                    // The number of bits needed to enumerate logical CPUs per core.
+                    u8::from(vcpu_config.vcpu_count > 1 && vcpu_config.smt),
+                )
+                .map_err(KvmVcpuConfigureError::NormalizeCpuidError)?;
         }
 
+        // Check support and set our CPUID.
+        let kvm_cpuid = {
+            // Get KVM supported CPUID.
+            let supported_cpuid = cpuid::Cpuid::kvm_get_supported_cpuid()
+                .map_err(KvmVcpuConfigureError::KvmGetSupportedCpuid)?;
+
+            // Checks KVM supports the given template.
```

Comment:
I understand your point here, although since we intend to implement user CPU templates in this feature branch I would learn towards maintaining this check in this PR. 

I could change:
```rust
// Check support and set our CPUID.
let kvm_cpuid = {
    // Get KVM supported CPUID.
    let supported_cpuid = cpuid::Cpuid::kvm_get_supported_cpuid()
        .map_err(KvmVcpuConfigureError::KvmGetSupportedCpuid)?;

    // Checks KVM supports the given template.
    supported_cpuid
        .supports(&config_cpuid)
        .map_err(KvmVcpuConfigureError::KvmUnsupportedCpuid)?;

    // Set CPUID.
    kvm_bindings::CpuId::from(config_cpuid)
};
```
to 
```rust
// Set CPUID.
let kvm_cpuid = kvm_bindings::CpuId::from(config_cpuid);
```
and then @mattschlebusch would need to include this change in his PR.

Reply:
Follow-up PR https://github.com/firecracker-microvm/firecracker/pull/3360